os/arch/arm/src/amebasmart/amebasmart_serial.c

```diff
@@ -1131,6 +1166,178 @@ static bool rtl8730e_up_txempty(struct uart_dev_s *dev)
+			pmu_register_sleep_callback(PMU_UART1_DEVICE, (PSM_HOOK_FUN)rtk_uart_suspend, NULL, (PSM_HOOK_FUN)rtk_uart_resume, NULL);
+			break;
+		default:
+			break;
+	}
+
+	return;
+}
+#endif
+
+/****************************************************************************
+ * Name: amebasmart_serial_pmprepare (DUMMY)
+ *
+ * Description:
+ *   Request the driver to prepare for a new power state. This is a warning
+ *   that the system is about to enter into a new power state. The driver
+ *   should begin whatever operations that may be required to enter power
+ *   state. The driver may abort the state change mode by returning a
+ *   non-zero value from the callback function.
+ *
+ * Input Parameters:
+ *
+ *    cb - Returned to the driver. The driver version of the callback
+ *         structure may include additional, driver-specific state data at
+ *         the end of the structure.
+ *
+ *    pmstate - Identifies the new PM state
+ *
+ * Returned Value:
+ *   Zero - (OK) means the event was successfully processed and that the
+ *          driver is prepared for the PM state change.
+ *
+ *   Non-zero - means that the driver is not prepared to perform the tasks
+ *              needed achieve this power setting and will cause the state
+ *              change to be aborted. NOTE: The prepare() method will also
+ *              be called when reverting from lower back to higher power
+ *              consumption modes (say because another driver refused a
+ *              lower power state change). Drivers are not permitted to
+ *              return non-zero values when reverting back to higher power
+ *              consumption modes!
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_PM
+static int amebasmart_serial_pmprepare(FAR struct pm_callback_s *cb, int domain,
+                                   enum pm_state_e pmstate)
+{
+	switch (pmstate)
+	{
+		case PM_NORMAL:
+		case PM_IDLE:
+		case PM_STANDBY:
+			break;
+
+		case PM_SLEEP:
+			/* UART0_DEV: hp uart0
+			 * UART1_DEV: hp uart1
+			 * UART2_DEV: hp uart2
+			 * UART3_DEV: hp uart3_bt
+			 * LOGUART_DEV: KM0 log uart */
+#ifdef CONFIG_RTL8730E_UART0
+			if ((g_uart0priv.txint_enable) && (!serial_writable(sdrv[0]))) {		/* If Tx init enable and FIFO not empty */
+					return ERROR;
+			}
+#endif
+#ifdef CONFIG_RTL8730E_UART1
+			if ((g_uart1priv.txint_enable) && (!serial_writable(sdrv[1]))) {		/* If Tx init enable and FIFO not empty */
+					return ERROR;
+			}
+#endif
+#ifdef CONFIG_RTL8730E_UART2
+			if ((g_uart2priv.txint_enable) && (!serial_writable(sdrv[2]))) {		/* If Tx init enable and FIFO not empty */
+					return ERROR;
+			}
+#endif
+#ifdef CONFIG_RTL8730E_UART4
+			// LOGUART_TypeDef *UARTLOG = LOGUART_DEV;
+			// if (!(UARTLOG->LOGUART_UART_LSR & LOG_UART_IDX_FLAG[2].empty)) {		/* log UART Tx Empty Check */
+			// 	return ERROR;
+			// }
```

Comment:
If these lines are not needed, can they be removed?

Reply:
Also @edwakuwaku , I remember that during our testing, UART4 would always reject PM state change because of some reason, is that resolved in this PR

Reply:
Yes, when we checked and tested on this part previously, the logic is correct as there is log need to be printed, so the condition for checking the bit will cause it not able to enter sleep. Thus, these lines should be removed.